src/proof.rs

```diff
@@ -0,0 +1,239 @@
+// Copyright 2020 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::nibbleslice::NibbleSlice;
+use crate::node::Node;
+use crate::{Trie, TrieDB, TrieError};
+use ccrypto::{blake256, BLAKE_NULL_RLP};
+use cdb::HashDB;
+use primitives::H256;
+
+// Unit of proof.
+//#[derive(Clone, Eq, PartialEq, Debug, RlpEncodable, RlpDecodable)]
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct CryptoProofUnit<H, K, V> {
+    hash: H, // or root
+    key: K,
+    value: V, // ignored in case of absence
+    presence: bool,
+}
+
+// Abstract trait of being cryptographically provable.
+pub trait CryptoProvable<H, K, V> {
+    // Does this verify the unit? Note: Invalid / ill-formed proof should return false, not abort or error
+    fn verify(&self, test: &CryptoProofUnit<H, K, V>) -> bool;
+}
+
+// Abstract trait of being a data structure that can provide CryptoProvable for its element.
+pub trait CryptoStructure<H, K, V> {
+    fn make_proof(&self, key: &K) -> crate::Result<(CryptoProofUnit<H, K, V>, Box<dyn CryptoProvable<H, K, V>>)>;
+}
+
+// This is strongly bound to implementation of TrieDB. I want to specify that but rust has no HKT(?)
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct CryptoProof_MerkleTrie {
+    proof: Vec<Vec<u8>>, // Starts with the closest node to root.
+}
+
+pub struct CryptoProof_SkewedTree {
+	// TODO
+}
+
+
+impl<'db> CryptoStructure<H256, H256, Vec<u8>> for TrieDB<'db> {
+	fn make_proof<'k>(&self, key: &'k H256) 
+	-> crate::Result<(CryptoProofUnit<H256, H256, Vec<u8>>, Box<dyn CryptoProvable<H256, H256, Vec<u8>>>)> {
+		type Tunit = CryptoProofUnit<H256, H256, Vec<u8>>;
+		type Tunit_partial = (H256, NibbleSlice<'k>, Vec<u8>, bool);
+
+		fn make_proof_upto(self_: &TrieDB<'_>, path: &NibbleSlice<'_>, hash: &H256) 
+		-> crate::Result<(Tunit_partial, Vec<Vec<u8>>)> {
+			let node_rlp = self_.db.get(&hash).ok_or_else(|| TrieError::IncompleteDatabase(*hash))?;
+
+			match Node::decoded(&node_rlp) {
+				Some(Node::Leaf(partial, value)) => {
+					if &partial == path {
+						Ok((hash, path, value, true))
```

Comment:
I think it's better to define a structure instead of using a tuple.

Reply:
Why?

Reply:
To name fields helps to read code.

Reply:
Is it common to define a temporary struct that is used internally only in a single function?

Reply:
Yes.

Reply:
Ok.